src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
+The parser runs the type checks and inference, so the information we get at this stage is already complete (explicit concrete monomorphic types, reified code only), and checked for syntactic and semantic errors.
 
-The MLIR dialect that we'll develop at this stage will be solely for representing the lowering concepts in order to simplify high-level optimisations before lowering to LLVM IR.
-In theory, the AST could have been directly lowered to LLVM IR, but the LLVM pipeline wouldn't be able, for example, to optimise based on type and region information, memory management mechanisms (ex. ref count), object layout and representation, etc.
+Verona also doesn't restrict declaration order. Functions and classes can be declared at any time in the module to be used anywhere.
+To avoid order issues in the IR, we'll do multiple passes on the structures:
+1. Declaration pass: Defines reachable type structures and functions, including the entire standard library files.
+2. Lowering pass: Lower the reachable AST nodes using the structures above and defining the functions for the prototypes above.
+
+In theory, the AST could have been directly lowered to LLVM IR, but the LLVM IR is too low level for some analysis we want to do (ex. region aliasing) and it would be counter-productive to do so at that stage.
+
+## Classes & Modules
+
+Classes and modules are the same concept in Verona.
+Each source sub-directory containing Verona files is a module, and all its Verona files share the same namespace, named after the directory name.
+The base source directory is also a module, and its sub-directories will be included in the "path" for importing modules.
+
+Calling for a method or accessing a field on an object of a type in a different module needs full scope.
+The parser will resolve all those issues and will cannonicalise the representation for all access to be fully scoped.
+
+Class types will be represented in a mix of LLVM structure types (for fields) and a `vtable` (as an array of pointers) for the dynamic methods.
+
+## Functions
+
+Functions are a native concept in MLIR and LLVM IR, and lowering free and static functions and calls to them is trivial.
+We'll declare all reachable functions beforehand, then lower the AST (implementing the prototypes as full functions).
+Dynamic resolution and selector colouring will be done at the parser level, and calls to dynamic methods in the AST will be a node with base pointer and offset.
+
+MLIR modules have a symbol table for functions, which we use to match declarations against definitions and calls.
+
+Arguments by value or reference have ABI reserved space (registers, stack), so we don't need to allocate them again (the `alloca` _trick_).
+Return values, for functions that do return a type, will be the last expression evaluated in the block.
+
+Lambdas will be lowered as an anonymous structure with captures as fields and `apply()` as the method for when the lambda is called.
```

Comment:
This lowering causes a lot of problems in clang.  In particular, it makes it incredibly hard to propagate information about aliasing state or mutability into the callee.  From the callee's perspective, all of the fields of the structure are just fields of a struct and so there's no way of propagating anything that we inferred about the variables that they capture into the lambda body or any clean way of running analyses that infer things across all uses.

It also makes some transforms tricky.  For example, if we discover after some inlining that one capture is always a constant, then we want to be able to remove it from the structure and constant propagate it into the lambda body.  This is very hard to do with the early lowering.  

There's ongoing work in clang to reverse this design decision there, but it's a multi-year engineering effort to undo after already going down the early-lowering path.  I was hoping that we'd be able to avoid this in Verona and keep lambdas abstract until LLVM IR generation.